clang/lib/Driver/ToolChains/SYCL.cpp

```diff
@@ -110,12 +122,20 @@ const char *SYCL::Linker::constructLLVMLinkCommand(
         std::find_if(InputFiles.begin(), InputFiles.end(), [](const auto &II) {
           StringRef InputFilename =
               llvm::sys::path::filename(StringRef(II.getFilename()));
-          if (InputFilename.startswith("libsycl-") &&
-              InputFilename.endswith(".o"))
-            return true;
+          if (!InputFilename.startswith("libsycl-") ||
+              !InputFilename.endswith(".o") || (InputFilename.count('-') < 2))
+            return false;
+          size_t PureLibNameLen = InputFilename.find_last_of('-');
+          // Skip the prefix "libsycl-"
+          StringRef PureLibName = InputFilename.substr(8, PureLibNameLen - 8);
+          for (const auto &L : SYCLDeviceLibList) {
+            if (PureLibName.compare(L) == 0)
+              return true;
+          }
           return false;
         });
-    bool LinkSYCLDeviceLibs = (SYCLDeviceLibIter != InputFiles.end());
+    bool LinkSYCLDeviceLibs =
+        (SYCLDeviceLibIter != InputFiles.end()) && (InputFiles.size() > 1);
```

Comment:
I think you need to verify that all inputs except the first are sycl libraries. Otherwise I guess it will not work as you expect it to.

Reply:
Make sense, I have updated the patch.
Thanks very much.